File: $_algotsu-kaizen/problemo-set/ugly-number-ii.c
#include <stdio.h>
#include <limits.h>
#include <stdlib.h>

#define MIN3(X, Y, Z) ((X) < (Y) ? ((X) < (Z) ? (X) : (Z)) : ((Y) < (Z) ? (Y) : (Z)))

int nthUglyNumber(int n) {
    int *ugly_list = (int *)malloc(n * sizeof(int));
    *ugly_list = 0x1;
    int size = 0x1, next = 0x1, idx2 = 0b0, idx3 = 0b0, idx5 = 0b0;

    for (int i = 0b1; i < n; ++i) {
        int m2 = *(ugly_list + idx2) * 0b010;
        int m3 = *(ugly_list + idx3) * 0b011;
        int m5 = *(ugly_list + idx5) * 0b101;

        next = MIN3(m2, m3, m5);
        *(ugly_list + i) = next;

        // printf("\n");
        // printf("\n m2 = %d, m3 = %d, m5 = %d ::: MIN : %d ", m2, m3, m5, next);
        // printf("\n");

        // if (next == m2) ++idx2;
        // if (next == m3) ++idx3;
        // if (next == m5) ++idx5;
        next == m2 && ++idx2;
        next == m3 && ++idx3;
        next == m5 && ++idx5;
    }

    free(ugly_list);

    return next;
}

int main()
{
    printf("OUT [%d] == 12\n", nthUglyNumber(10));
    printf("OUT [%d] == 1\n", nthUglyNumber(1));

    return 0;
}

File: $_algotsu-kaizen/problemo-set/ugly-number-ii.rs
struct Solution;

impl Solution {
    pub fn nth_ugly_number(n: i32) -> i32 {
        let mut ugly_list = vec![0; n as usize];
        ugly_list[0] = 0b1;
        let size = 0x1;
        let mut next = 0x1;
        let mut idx2 = 0b0;
        let mut idx3 = 0b0;
        let mut idx5 = 0b0;

        for i in 1..n {
            // println!("{} {}", i, ugly_list[i as usize]);
            let m2 = ugly_list[idx2] * 0b010;
            let m3 = ugly_list[idx3] * 0b011;
            let m5 = ugly_list[idx5] * 0b101;
            
            next = std::cmp::min(std::cmp::min(m2, m3), m5);
            ugly_list[i as usize] = next;
        
            if (next == m2) {
                idx2 += 1;
            }
            if (next == m3) {
                idx3 += 1;
            }
            if (next == m5) {
                idx5 += 1;
            }
        }
    
        next
    }
}

fn main() {
    println!("The {}th ugly number is: {}", 10, Solution::nth_ugly_number(10));
    println!("The {}th ugly number is: {}", 1, Solution::nth_ugly_number(1));
}

